bundles/cif-virtual-catalog/src/main/java/com/adobe/cq/commerce/virtual/catalog/data/impl/CatalogDataResourceProviderManagerImpl.java

```diff
@@ -174,8 +173,10 @@
                 log.error("Error finding data roots", x);
             }
         }
-        dataRoots = allResources;
-        return allResources;
+        synchronized (dataRoots) {
+            dataRoots.clear();
+            dataRoots.addAll(allResources);
+        }
     }
 
     private void registerDataRoots() {
```

Comment:
Instead of the fine-granular concurrent access control, I think its is easier to just make the registerDataRoots() and registerDataRoot() synchronised.